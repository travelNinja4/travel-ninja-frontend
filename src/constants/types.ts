/**
 * used to store global types in the project
 *
 */

export type OtpProps = {
  otpCode: string;
  otpType?: 'email' | 'mobile';
};

export interface LoginResponse {
  data: {
    phoneNumber?: string;
  };
}
