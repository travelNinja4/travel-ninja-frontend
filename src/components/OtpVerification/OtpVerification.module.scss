/**
------------------------------------------------------
styles specific to the OtpVerification component
------------------------------------------------------
**/
@use '../../styles/mixins' as *;

.container {
  position: relative;
  display: flex;
  min-height: 100vh;
  overflow: hidden;
  padding: 1rem;
  align-items: center;
  justify-content: center;

  .bgImg {
    object-fit: cover;
    z-index: 0;
  }

  .overlay {
    position: absolute;
    inset: 0;
    background: linear-gradient(135deg, var(--color-indigo) 0%, var(--color-purple) 100%);
    z-index: 1;
    opacity: 0.8;
  }

  .subContainer {
    background: var(--color-white);
    border-radius: 20px;
    box-shadow: 0 20px 40px var(--color-black-alpha-10);
    overflow: hidden;
    min-height: 600px;
    max-width: 1000px;
    margin: 0 auto;
    height: auto;
    position: relative;
    z-index: 1;

    .mainContainer {
      display: flex;
      flex: 1;

      flex-direction: column;

      .userData {
        align-items: center;
        justify-content: center;
        background-color: var(--color-gray-100);
        border: 1px solid var(--color-gray-150);
        padding: 1rem;
        margin-bottom: 1.5rem;
        border-radius: 8px;
        display: flex;
        flex-direction: row;

        .userEmail {
          padding-left: 6px;
        }
      }

      .imageContainer {
        position: relative;
        display: flex;
        align-items: center;
        justify-content: center;
        flex: 0.5;
        min-height: auto;
      }

      .otpContainer {
        display: flex;
        flex-direction: column;
        flex: 0.5;
        background-color: var(--color-white);
        padding: 1rem;
      }

      .verificationWrapper {
        margin-bottom: 1.5rem;
      }

      .verificationTextField {
        text-align: center;
        font-size: 1.5rem;
        font-weight: 600;
        letter-spacing: 0.5rem;
        padding: 1rem;
      }

      .button {
        color: var(--color-white);
      }

      .resendWrapper {
        margin-top: 1rem;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;
      }

      .resendTextContainer {
        display: flex;
        flex: 1;
        align-items: center;
        justify-content: center;
        gap: 8px;
        cursor: pointer;
      }

      .resendText {
        font-size: 16px;
        font-weight: 500;
        color: var(--color-indigo);
      }

      @include respond-to(md) {
        .imageContainer {
          min-height: 100%;
          height: auto;
        }
      }

      @include respond-to(lg) {
        flex-direction: row;

        .otpContainer {
          padding: 2rem;
        }

        .otpHeader {
          font-weight: 700;
          margin-bottom: 0.5rem;
        }

        .otpSubHeader {
          color: var(--color-gray-900-alpha-75);
          margin-bottom: 1.5rem;
        }

        .imageContainer {
          flex: 0.5;
          min-height: 600px;
        }

        .otpContainer {
          flex: 0.5;
        }
      }
    }
  }
}
