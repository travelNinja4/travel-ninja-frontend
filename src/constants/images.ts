/**
 * Handles all the cloudinary image constants
 *
 */

export const IMAGES = {
  FALLBACK_IMG: 'fallbackImg.png',
};
