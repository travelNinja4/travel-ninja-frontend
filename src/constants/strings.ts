export const STRINGS = {
  APP_NAME: 'TravelNinja',
  CREATE_ACCOUNT_APP_TAG_LINE: 'Transform your tour business with our powerful platform',
  LOGIN_APP_TAG_LINE: 'Empowering tour agencies to create unforgettable travel experiences',
  RESET_PASSWORD_TAG_LINE: 'Reset your password to regain access to your account',
  EMAIL: 'email',
  MOBILE: 'mobile',
  EMAIL_VERIFICATION: 'Email Verification',
  MOBILE_NUMBER_VERIFICATION: 'Mobile Number Verification',
  EMAIL_SENT: `We've sent a verification code to your email`,
  MOBILE_SENT: `We've sent a verification code to your mobile number`,
  EMAIL_CODE_HINT: 'Enter the 6-digit code sent to your email',
  MOBILE_CODE_HINT: 'Enter the 6-digit code sent to your mobile number',
  VERIFY_EMAIL: 'Verify Email',
  VERIFY_MOBILE_NUMBER: 'Verify Mobile Number',
  ENTER_VERIFICATION_CODE: 'Enter Verification Code',
  RESEND_CODE_IN: 'Resend code in',
  SECONDS: 'seconds',
  RESEND_CODE: 'Resend Code',
  NO_OPTIONS_FOUND: 'No options found',
  COUNTRY_CODE: 'Country code',
  ENTER_MOBILE_NUMBER: 'Enter mobile number',
  FORGOT_PASSWORD: 'Forgot Password?',
  RESET_PASSWORD_INSTRUCTION: 'Enter your email to receive reset instructions',
  RESET_PASSWORD_LABEL: `We'll send you a link to reset your password`,
  SEND_RESET_LINK: 'Send Reset Link',
  EMAIL_ADDRESS: 'Email Address',
  ENTER_EMAIL_ADDRESS: 'Enter your email address',
  BACK_TO_LOGIN: 'Back to Login',
};

export const ROUTES = {
  HOME: '/',
  LOGIN: '/login',
  CREATE_ACCOUNT: '/createAccount',
  VERIFY: '/verify',
} as const;

export const NOTIFICATION_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
  INFO: 'info',
};
