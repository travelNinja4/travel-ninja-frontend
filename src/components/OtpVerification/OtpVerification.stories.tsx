import type { Meta, StoryObj } from '@storybook/nextjs-vite';
/**
import { fn } from 'storybook/test';
*/
import OtpVerification from './OtpVerification';

const meta = {
  title: 'components/OtpVerification',
  component: OtpVerification,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {},
  args: {
    /** onClick: fn() **/
  },
} satisfies Meta<typeof OtpVerification>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
