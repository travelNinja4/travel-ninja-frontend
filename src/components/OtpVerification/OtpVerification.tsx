/**
 * A reusable component for verifying users via a one-time password (OTP) sent to their email or mobile number, with support for code input, verification, and resending OTP.
 *
 * @example
 * ```tsx
 * import OtpVerification from '@src/components/OtpVerification'
 *
 * export default function OtpVerification() {
 *   return <OtpVerification label="Hello" />;
 * }
 * ```
 */
'use client';

import { useEffect, useState } from 'react';
import { ROUTES, STRINGS } from '@/constants/strings';
import AuthSideBanner from '../AuthSideBanner';
import Button from '../Button';
import TextField from '../TextField';
import Typography from '../Typography';
import CustomImage from '../CustomImage';
import { Mail, Phone, Check, RotateCw } from 'lucide-react';
import { useAuthStore } from '@/store/auth';
import { useRouter } from 'next/navigation';
import { authService } from '@/services/authService';
import styles from './OtpVerification.module.scss';

export default function OtpVerification() {
  const router = useRouter();
  const accountData = useAuthStore((store) => store.accountData);
  const [type, setType] = useState<'email' | 'mobile'>('email');
  const [value, setValue] = useState('');
  const [timer, setTimer] = useState(0);
  const [verifyLoading, setVerifyLoading] = useState(false);

  useEffect(() => {
    if (timer === 0) return;

    let interval: NodeJS.Timeout | null = null;
    if (timer > 0) {
      interval = setInterval(() => setTimer((prev) => prev - 1), 1000);
    }
    return () => {
      if (interval) clearInterval(interval);
    };
  }, [timer]);

  const handleResend = async () => {
    try {
      const requestParams =
        type === STRINGS.EMAIL
          ? { email: accountData?.email }
          : { phoneNumber: `+91${accountData?.phoneNumber}` };
      console.log('requestParams>>>', requestParams);
      const response = await authService.sendOtp(requestParams);
      console.log('Otp sent successfully:', response);
      setTimer(60);
    } catch (err: unknown) {
      console.log('error>>>', err);
    }
  };

  const handleOnChange = (val: string) => {
    setValue(val);
  };

  const handleCodeVerify = async () => {
    if (!value) {
      return false;
    }

    try {
      const requestParams =
        type === STRINGS.EMAIL
          ? { email: accountData?.email, otpCode: value }
          : { phoneNumber: `+91${accountData?.phoneNumber}`, otpCode: value };
      console.log('requestParams>>>', requestParams);
      setVerifyLoading(true);
      const response = await authService.verifyOtp(requestParams);
      console.log('Otp sent successfully:', response);
      if (type === STRINGS.EMAIL) {
        setType('mobile');
        setValue('');
        setTimer(0);
      } else {
        router.push(ROUTES.LOGIN);
      }
    } catch (err: unknown) {
      console.log('error>>>', err);
    } finally {
      setVerifyLoading(false);
    }
  };

  return (
    <div data-testid="OtpVerificationTest" className={styles.container}>
      <CustomImage
        isUrl
        src={process.env.NEXT_PUBLIC_PLACEHOLDER_BG_URL as string}
        fill
        priority
        alt="BackgroundImg"
        className={styles.bgImg}
        unoptimized
      />
      <div className={styles.overlay}></div>

      <div className={styles.subContainer}>
        <div className={styles.mainContainer}>
          <div className={styles.imageContainer}>
            <AuthSideBanner appTagLine={STRINGS.CREATE_ACCOUNT_APP_TAG_LINE} />
          </div>
          <div className={styles.otpContainer}>
            <Typography tag="h2" align="center" className={styles.otpHeader}>
              {type === STRINGS.EMAIL
                ? STRINGS.EMAIL_VERIFICATION
                : STRINGS.MOBILE_NUMBER_VERIFICATION}
            </Typography>
            <Typography tag="p" align="center" className={styles.otpSubHeader}>
              {type === STRINGS.EMAIL ? STRINGS.EMAIL_SENT : STRINGS.MOBILE_SENT}
            </Typography>
            <div className={styles.userData}>
              {type === STRINGS.EMAIL ? (
                <Mail color="var(--color-black)" />
              ) : (
                <Phone color="var(--color-black)" />
              )}
              <Typography
                tag="span"
                align="center"
                color="var(--color-indigo)"
                className={styles.userEmail}
              >
                {type === STRINGS.EMAIL ? accountData?.email : accountData?.phoneNumber}
              </Typography>
            </div>
            <div className={styles.verificationWrapper}>
              <TextField
                label={STRINGS.ENTER_VERIFICATION_CODE}
                placeholder="000000"
                value={value}
                maxLength={6}
                className={styles.verificationTextField}
                onChange={(e) => handleOnChange(e.target.value)}
              />
              <Typography tag="label" align="center" className={styles.otpSubHeader}>
                {type === STRINGS.EMAIL ? STRINGS.EMAIL_CODE_HINT : STRINGS.MOBILE_CODE_HINT}
              </Typography>
            </div>
            <Button
              className={styles.button}
              startIcon={Check}
              onClick={handleCodeVerify}
              disabled={verifyLoading}
              loading={verifyLoading}
            >
              {type === STRINGS.EMAIL ? STRINGS.VERIFY_EMAIL : STRINGS.VERIFY_MOBILE_NUMBER}
            </Button>
            <div className={styles.resendWrapper}>
              {timer > 0 ? (
                <Typography tag="span">{`${STRINGS.RESEND_CODE_IN} ${timer} ${STRINGS.SECONDS}`}</Typography>
              ) : (
                <>
                  <RotateCw color="var(--color-indigo)" />
                  <Typography tag="span" className={styles.resendText} onClick={handleResend}>
                    {STRINGS.RESEND_CODE}
                  </Typography>
                </>
              )}
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
