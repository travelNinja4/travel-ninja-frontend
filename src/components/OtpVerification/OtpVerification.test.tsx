import '@testing-library/jest-dom';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { STRINGS } from '@/constants/strings';
import OtpVerification from './OtpVerification';

describe('Home', () => {
  /** Base props for OtpVerification **/
  const baseProps = {};

  /**
   * @test
   * Renders the OtpVerification component without throwing an error
   * and checks if the OtpVerificationTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<OtpVerification {...baseProps} />);
    expect(screen.getByTestId('OtpVerificationTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the OtpVerification component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<OtpVerification {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the OtpVerification component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<OtpVerification {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls handleResend and starts timer', () => {
    jest.useFakeTimers();
    render(<OtpVerification {...baseProps} />);
    fireEvent.click(screen.getByText(/Resend Code/i));

    expect(screen.getByText(`${STRINGS.RESEND_CODE} 60 ${STRINGS.SECONDS}`)).toBeInTheDocument();

    // Fast-forward 1 second
    act(() => {
      jest.advanceTimersByTime(1000);
    });

    expect(screen.getByText(`${STRINGS.RESEND_CODE} 59 ${STRINGS.SECONDS}`)).toBeInTheDocument();
    jest.useRealTimers();
  });

  it('updates value when code is typed', () => {
    render(<OtpVerification {...baseProps} />);
    const input = screen.getByPlaceholderText('000000');
    fireEvent.change(input, { target: { value: '123456' } });
    expect(input).toHaveValue('123456');
  });

  it('does nothing if verify clicked with empty value', () => {
    render(<OtpVerification {...baseProps} />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(screen.getByText(STRINGS.EMAIL_VERIFICATION)).toBeInTheDocument();
  });

  it('switches from email to mobile on verify with value', () => {
    render(<OtpVerification {...baseProps} />);
    const input = screen.getByPlaceholderText('000000');
    fireEvent.change(input, { target: { value: '123456' } });

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(screen.getByText(STRINGS.MOBILE_NUMBER_VERIFICATION)).toBeInTheDocument();
  });
});
