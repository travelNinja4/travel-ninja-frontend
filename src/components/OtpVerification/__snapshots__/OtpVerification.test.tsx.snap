// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="OtpVerificationTest"
  >
    <img
      alt="BackgroundImg"
      blurdataurl="https://res.cloudinary.com/dyci0wg0i/image/upload/fallbackImg.png"
      class="bgImg"
      data-testid="CustomImageTest"
      placeholder="blur"
      src="https://picsum.photos/1920/1080?random=1"
    />
    <div
      class="overlay"
    />
    <div
      class="subContainer"
    >
      <div
        class="mainContainer"
      >
        <div
          class="imageContainer"
        >
          <div
            class="container"
            data-testid="AuthSideBannerTest"
          >
            <img
              alt="authImg"
              blurdataurl="https://res.cloudinary.com/dyci0wg0i/image/upload/fallbackImg.png"
              class="imgContainer"
              data-testid="CustomImageTest"
              placeholder="blur"
              src="https://picsum.photos/1920/1080?random=1"
            />
            <div
              class="overlay"
            />
            <div
              class="headerContainer"
            >
              <h1
                class="typography h1 align-center header"
                data-testid="TypographyTest"
                style="color: var(--color-white);"
              >
                TravelNinja
              </h1>
              <p
                class="typography p align-center subHeader"
                data-testid="TypographyTest"
                style="color: var(--color-white);"
              >
                Transform your tour business with our powerful platform
              </p>
            </div>
          </div>
        </div>
        <div
          class="otpContainer"
        >
          <h2
            class="typography h2 align-center otpHeader"
            data-testid="TypographyTest"
          >
            Mobile Number Verification
          </h2>
          <p
            class="typography p align-center otpSubHeader"
            data-testid="TypographyTest"
          >
            We've sent a verification code to your mobile number
          </p>
          <div
            class="userData"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-phone"
              fill="none"
              height="24"
              stroke="var(--color-black)"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.832 16.568a1 1 0 0 0 1.213-.303l.355-.465A2 2 0 0 1 17 15h3a2 2 0 0 1 2 2v3a2 2 0 0 1-2 2A18 18 0 0 1 2 4a2 2 0 0 1 2-2h3a2 2 0 0 1 2 2v3a2 2 0 0 1-.8 1.6l-.468.351a1 1 0 0 0-.292 1.233 14 14 0 0 0 6.392 6.384"
              />
            </svg>
            <span
              class="typography span align-center userEmail"
              data-testid="TypographyTest"
              style="color: var(--color-indigo);"
            >
              US-1234567890
            </span>
          </div>
          <div
            class="verificationWrapper"
          >
            <div
              class="container"
              data-testid="TextFieldTest"
            >
              <label
                class="typography label align-left label"
                data-testid="TypographyTest"
              >
                Enter Verification Code 
              </label>
              <div
                class="inputWrapper"
              >
                <input
                  class="inputField verificationTextField"
                  maxlength="6"
                  placeholder="000000"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <label
              class="typography label align-center otpSubHeader"
              data-testid="TypographyTest"
            >
              Enter the 6-digit code sent to your mobile number
            </label>
          </div>
          <button
            class="button button"
            data-testid="ButtonTest"
            type="button"
          >
            <span
              class="icon"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-check"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 6 9 17l-5-5"
                />
              </svg>
            </span>
            <span
              class="typography span align-center button"
              data-testid="TypographyTest"
            >
              Verify Mobile Number
            </span>
          </button>
          <div
            class="resendWrapper"
          >
            <div
              class="resendTextContainer"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-rotate-cw"
                fill="none"
                height="24"
                stroke="var(--color-indigo)"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21 12a9 9 0 1 1-9-9c2.52 0 4.93 1 6.74 2.74L21 8"
                />
                <path
                  d="M21 3v5h-5"
                />
              </svg>
              <span
                class="typography span align-left resendText"
                data-testid="TypographyTest"
              >
                Resend Code
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
